### PR TITLE
feat(memory): ship confirmed embedding rebuild recovery

### LINKED ISSUE #1314: feat(memory): runtime rebuild operation with confirmed embedding change
## Problem Statement

Changing `processing.embedding.base_url` or `processing.embedding.model` can leave Memory in a durable dead end. Avibe saves an `embedding_change_pending` marker, but reconcile only settles that marker when every vector-bearing surface is empty. With existing data the save fails, the UI rolls the candidate configuration back, and Restart engine remains fenced. The only documented escape today is Clear all, which deletes the user's memory.

The managed-child prerequisites are already merged into `dev`: artifact compatibility admission (#1307) and the role-aware `cascade rebuild` child capability (#1310). What is missing is the user-reachable recovery path: confirmed settings flow, runtime-owned rebuild orchestration over that child, an authenticated route, and the minimal Settings UI needed to confirm/retry.

## Solution

Ship a **narrow rebuild recovery path**, not a generic operations framework:

1. Confirmed embedding identity change on Memory settings PATCH.
2. Durable candidate + pending marker before any destructive work.
3. A **cross-process Memory config transaction** so UI save and Controller settlement cannot clobber each other with stale snapshots.
4. `MemoryRuntime.rebuild()` as one retained/shielded task (same shape as existing restart): the HTTP request **waits for the final result**.
5. Minimal Settings UI for confirm / retry / `rebuild_required` warning.

Markdown is preserved. Old and new vector spaces are never mixed. Failure leaves a truthful Retry path and never silently escalates to factory reset.

## User Stories

1. As an Avibe user, I want to change embedding provider/model without losing markdown memory content.
2. As an Avibe user, I want an unconfirmed identity-changing save rejected with a clear rebuild-required error.
3. As an Avibe user, I want the confirmed candidate saved durably before destructive work so a crash retries with the same config.
4. As an Avibe user, I want a failed rebuild to keep the confirmed configuration (no post-confirm rollback).
5. As an Avibe user, I want an empty root to settle without launching a child.
6. As an Avibe user, I want an API-key correction under a pending marker to feed the next retry.
7. As an Avibe user, I want API-key-only changes with no pending marker to keep ordinary save behavior.
8. As an Avibe user, I want concurrent Memory mutations refused with a distinct in-progress error while rebuild runs.
9. As an Avibe user, I want a duplicate rebuild request to join the running task rather than fork.
10. As an Avibe user, I want boot with a pending marker to keep claims fenced and the sidecar down, with Retry exposed and no automatic destructive work.
11. As an Avibe user, I want root-busy reported via a stable public error code and never steered toward factory reset.
12. As an operator, I want status to expose only a minimal rebuild failure/pending projection — no secrets, job IDs, history, or progress endpoints.

### Confirmed settings write

- `PATCH /api/memory/settings` remains the only candidate-configuration write surface.
- Optional exact boolean `confirm_rebuild`.
- Unconfirmed identity change → `409 memory_embedding_rebuild_required`; **nothing saved** (including apparently empty roots, to avoid check-then-save races).
- Confirmed save persists candidate + pending marker **before** scheduling rebuild. No rollback after confirmation.
- Settings GET exposes read-only `rebuild_required` while the marker is pending.
- API-key-only, no marker: ordinary save/reconcile/rollback.
- API-key-only under marker: update candidate, return saved-but-`rebuild_required`, do not reconcile or clear the fence.

### Cross-process Memory config transaction (required)

UI and Controller run in different processes. Process-local UI locks and `CONFIG_LOCK` are **not** a cross-process transaction. Without an explicit narrow protocol, a concurrent Controller write-back can overwrite a freshly saved candidate or clear the marker.

Define a **narrow Memory config transaction**, not a general multi-writer config system:

- Candidate + pending marker are written as one durable unit visible to both processes.
- Controller settlement (clear marker / activate candidate) and UI save of a new candidate must not lose the last confirmed write.
- Stale Controller snapshots must not clobber a newer confirmed candidate/marker.
- After a successful rebuild settles the marker and **before** sidecar activation, update the Controller/restart candidate snapshot so activation uses the settled config.
- Cover with an explicit **stale save / rollback interleaving** test at the routes + runtime boundary (hermetic; no live service).

Do **not** invent a generic distributed lock service or operation journal.

### Rebuild orchestration

- Deep module is `MemoryRuntime`. Reuse the existing retained/shielded task + shutdown-join pattern from `restart()` (`_restart_task`).
- Add `MemoryRuntime.rebuild()` and `_rebuild_task`. Request waits for the final result (200 with closed outcome, or structured error). **No 202, no polling, no job IDs, no generic Controller operation projection, no reserved `factory_reset` kind.**
- At most one rebuild task; join-on-duplicate; other Memory mutations see `memory_operation_in_progress` (or the existing clear/maintenance conflict code if that is already the chokepoint — prefer one stable public code, document it in this issue’s acceptance).
- Before any destructive stop: validate pinned artifact usability and that the candidate embedding endpoint is complete enough to rebuild a non-empty root. Fail closed with no destructive work when validation fails.
- Order under existing reconcile/destructive lifecycle:
  1. claim fence + definitive quiesce (no force bypass)
  2. stop sidecar with proven process-group death
  3. strict-empty fast path → synthetic `completed_empty` (no child)
  4. otherwise launch pinned-artifact `cascade rebuild --yes` via the merged role-aware child with bounded timeout
  5. map closed child results; settle marker only on `completed` / `completed_empty`
  6. clear marker durably, refresh Controller/restart candidate snapshot, then destructive-cutover start (skip ordinary healthy-replacement endpoint preflight)
- Closed result set: `completed | completed_empty | root_busy | interrupted | timed_out | failed`.
- `root_busy` (EverOS exit 3) maps to a **stable public error code** (e.g. `memory_rebuild_root_busy`) and must never mention factory reset.
- Indeterminate data state fails closed.
- Boot with pending marker: reconcile ownership first, keep claims fenced, no sidecar, expose retry projection; never auto-start rebuild.
- Health after enabled startup: existing health projection only — operational warning vs failed-row data-quality warning; no stdout parsing.

### Routes

- `POST /api/memory/runtime/rebuild` accepts exactly `{"confirm": true}`; reuses existing CSRF / signed-user-proof chain; no config or force fields.
- Settings-driven confirmed path schedules the same `MemoryRuntime.rebuild()` after durable candidate+marker write.
- Manual “Rebuild index” without a settings change is **out of scope for this issue** (can follow once the marker/retry path is solid). Retry while the marker is already pending is in scope.

### Minimal Settings UI (ships with this issue)

Necessary UI only — not a horizontal recovery-UI bucket:

- Identity-changing Save opens a non-destructive confirmation (cost/duration disclosure); draft retained and same patch replayed with `confirm_rebuild: true`.
- Embedding identity fields remain editable when data exists (remove the data-exists lock).
- Persistent `rebuild_required` warning while marker pending.
- Footer: `Retry rebuild` while marker pending (beside existing Clear all). No permanent extra Rebuild control required beyond retry under pending marker.
- `Restart engine` disabled while rebuild required or rebuild task running.
- All new copy through `ui/src/i18n/en.json` and `zh.json`.

Processing Record state rows, dual-warning polish beyond what health already shows, Dependencies-page work, and full recovery-ladder user docs are **not** in this issue.

### Error codes (public, i18n-cataloged)

- `memory_embedding_rebuild_required`
- `memory_rebuild_failed`
- `memory_rebuild_root_busy`
- `memory_operation_in_progress` (or document reuse of an existing conflict code if one already gates all Memory mutations)

## Testing Decisions

Hermetic only: test-owned homes, fake artifacts/interpreters, no network, no local `vibe` service, no production paths.

1. **UI memory routes**: confirmed PATCH, 409 without confirm, exact rebuild body, CSRF/signed proof, join-on-duplicate, final-result responses (not 202), no secrets in payloads, cross-process/stale-save interleaving against a fake dual-writer or sequenced Controller write-back.
2. **MemoryRuntime layer**: quiesce failure, validation-before-destructive-stop, strict-empty vs indeterminate, result→settlement mapping, marker clear then candidate-snapshot refresh before activation, boot/orphan matrix, disabled-candidate empty settlement.
3. **Focused UI tests** where the existing frontend harness already supports them: draft retention + confirm replay, retry label under pending marker, restart disabled while pending.

Assert external behavior only (responses, persisted marker/candidate, child argv/env, process-group death, projection fields) — never internal call order or stdout.

Acceptance criteria in this issue are **self-contained**. Untracked companion plans under `docs/plans/` or `docs/memory-*.md` are background only and are **not** acceptance contracts. Do not require “plan test group N”.

## Out of Scope

- Factory reset (#1315).
- Generic Controller operation framework, 202 Accepted, client polling, operation IDs/history/progress, reserved `factory_reset` operation kind.
- Manual rebuild without a pending marker.
- Processing Record operation row, Dependencies-page UI, full recovery-ladder user documentation.
- Repair index / `cascade sync` / `fix --apply` (#1318).
- Release workflow / multi-platform artifact publish (already present in `.github/workflows/publish.yml`; not implementation scope here).
- Force restart, automatic rebuild from health, `cascade backfill`, stdout parsing, EverOS source changes, snapshot format changes.

## Dependencies

- Requires merged #1307 and #1310 on `dev`.
- Base PR on latest `origin/dev`.

### PR DESCRIPTION
**Pull request by rkrkrkk**
Closes #1314

## Summary
- Confirmed embedding identity changes on PATCH /api/memory/settings (confirm_rebuild) persist candidate + embedding_change_pending before destructive work; unconfirmed identity changes return 409 memory_embedding_rebuild_required with nothing saved.
- Narrow cross-process Memory config transaction (memory_config_transaction file lock + compare-and-swap) so UI saves and Controller settlement cannot clobber each other.
- MemoryRuntime.rebuild() retained _rebuild_task (join-on-duplicate; HTTP waits for final result). Validate artifact/endpoint before stop, fence/quiesce, stop sidecar, empty fast path or cascade rebuild child, settle only on success, refresh restart snapshot, then cutover start.
- Stable public errors: memory_embedding_rebuild_required, memory_rebuild_failed, memory_rebuild_root_busy, memory_operation_in_progress. Never steers to factory reset.
- Minimal Settings UI: confirm dialog + draft replay, editable embedding identity, rebuild_required warning, Retry rebuild, Restart engine disabled while pending/running. en+zh i18n.

## Evidence layers
- Unit: MemoryRuntime rebuild validation / empty settlement / root_busy / snapshot refresh; pending-marker fence on reconcile/restart.
- Contract: UI memory routes for confirm required, confirmed save+rebuild wait, failed rebuild keeps candidate, API-key-under-marker, exact rebuild body, stale-save CAS interleaving, ordinary non-identity save serialization.
- Scenario: none (no catalog entry for this recovery path).
- Manual residual: end-to-end embedding change against a non-empty root in Incus regression after merge.

## Dependencies
- Requires #1307 and #1310 (already on dev).

### CODE REVIEW

**Comment by cyhhao** [comment]
@codex review

File: vibe/ui_memory_routes.py [hunk 1]
```diff
@@ -261,15 +322,46 @@ async def _apply_memory_settings_patch(
...
+                {
+                    "status": "failed",
+                    "error": "memory_embedding_rebuild_required",
+                },
+                status_code=409,
+            )
+
+        # API-key (or other non-identity) updates under an existing marker update
+        # the candidate only. They do not reconcile, clear the fence, or roll back.
+        if pending_marker and not identity_changed:
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply vibe/ui_memory_routes.py]: Reconciled: enabled changes under a pending marker now go through the normal save+reconcile path so disable cannot leave a live runtime admitting captures.

File: ui/src/components/settings/memory/MemorySettingsPanel.tsx [hunk 2]
```diff
@@ -150,67 +153,57 @@ export const MemorySettingsPanel: React.FC<{
...
-      }
       const res = await api.saveMemorySettings(patch);
       if (isMemoryOk(res)) {
         onSaved(res);
-        showToast(t('memory.settings.saved'), 'success');
+        if (res.rebuild_required) {
+          const runtimeOk = (res.runtime as { ok?: boolean } | undefined)?.ok;
+          showToast(
+            runtimeOk === false ? t('memory.settings.rebuildFailed') : t('memory.settings.rebuildCompleted'),
+            runtimeOk === false ? 'error' : 'success',
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply ui/src/components/settings/memory/MemorySettingsPanel.tsx]: Fixed: toast now keys off runtime.ok when a rebuild actually ran; credential-only saves under a pending marker show the ordinary saved toast.

File: vibe/api.py [hunk 3]
```diff
@@ -1009,20 +1009,56 @@ def save_config(
...
+    """
 
     if not isinstance(memory_payload, dict):
         raise ValueError("Memory config payload must be an object")
     if not isinstance(embedding_change_pending, bool):
         raise ValueError("Memory embedding compatibility state must be a boolean")
     payload = dict(memory_payload)
     payload["embedding_change_pending"] = embedding_change_pending
-    return save_config({"memory": payload}, allow_memory=True)
+    with memory_config_transaction():
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply vibe/api.py:1091]: Fixed: save_config now shares memory_config_transaction and re-reads the latest Memory unit for non-memory writers so a stale whole-config snapshot cannot re-arm the marker.

File: vibe/ui_memory_routes.py [hunk 4]
```diff
@@ -279,15 +371,50 @@ async def _apply_memory_settings_patch(
...
+            payload = _settings_ok_payload(latest.memory, runtime_payload)
+            if status_code != 200 or runtime_payload.get("ok") is not True:
+                error = _memory_closed_error(
+                    runtime_payload,
+                    fallback="memory_rebuild_failed",
+                )
+                payload["status"] = "failed"
+                payload["error"] = error
+                # Confirmed candidate stays; surface rebuild_required for Retry.
+                payload["rebuild_required"] = True
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply vibe/ui_memory_routes.py]: Fixed: failed confirmed rebuild responses keep rebuild_required from the durable latest config projection rather than forcing true after settlement.

File: ui/src/components/settings/memory/MemorySettingsPanel.tsx [hunk 5]
```diff
@@ -150,67 +153,57 @@ export const MemorySettingsPanel: React.FC<{
...
+        }
+        setConfirmRebuildOpen(false);
+        setPendingPatch(null);
       } else {
         setError(memoryErrorMessage(t, (res as { error?: string })?.error));
-        // Reconciliation may roll back endpoint fields. Reload both resources
-        // instead of restoring drafts from the stale pre-save snapshot.
+        // Confirmed rebuild keeps the candidate on failure; reload both resources
+        // so Retry / rebuild_required reflect the durable marker.
         onReloadSettings();
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply ui/src/components/settings/memory/MemorySettingsPanel.tsx:219]: Fixed: confirm dialog and pending patch clear on failure so Retry rebuild is the recovery control under the durable marker.

**Comment by cyhhao** [comment]
@codex review

File: core/memory/runtime.py [hunk 6]
```diff
@@ -661,6 +663,9 @@ async def reconcile(self, config: MemoryConfig) -> dict[str, Any]:
     async def _reconcile(self, config: MemoryConfig) -> dict[str, Any]:
         """Run one reconciliation owned by the public lifecycle operation."""
 
+        if self._rebuild_running() and asyncio.current_task() is not self._rebuild_task:
+            return {"ok": False, "error": "memory_operation_in_progress"}
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply core/memory/runtime.py:667]: Fixed: after acquiring _reconcile_lock, rebuild ownership is rechecked and the durable Memory unit is preferred over a stale pending snapshot.

File: core/memory/runtime.py [hunk 7]
```diff
@@ -1633,21 +1699,207 @@ async def _restart_locked(self) -> dict[str, Any]:
...
+                child_result = RebuildProcessResult.COMPLETED
+                mapped = {
+                    "ok": True,
+                    "result": "completed_empty",
+                }
+
+            if not await asyncio.to_thread(
+                self._settle_embedding_change_pending,
+                candidate,
+            ):
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply core/memory/runtime.py:1828]: Fixed: settlement now compares embedding identity only and activation reloads the latest durable candidate so non-identity key/LLM fixes under the marker are preserved.

File: ui/src/components/settings/memory/MemorySettingsPanel.tsx [hunk 8]
```diff
@@ -150,81 +153,129 @@ export const MemorySettingsPanel: React.FC<{
...
+    const patch = buildPatch();
+    if (Object.keys(patch).length === 0) {
+      showToast(t('memory.settings.saved'), 'success');
+      return;
+    }
+    if (identityChanged(embeddingDraft, settings.processing.embedding)) {
+      // Retain the draft and open confirmation; the same patch is replayed with
+      // confirm_rebuild: true after the user accepts the cost/duration disclosure.
+      setPendingPatch(patch);
+      setConfirmRebuildOpen(true);
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply ui/src/components/settings/memory/MemorySettingsPanel.tsx:244]: Fixed: the UI identity check now matches the backend first-time-setup predicate and skips rebuild confirmation when both original fields are empty.

File: config/v2_config.py [hunk 9]
```diff
@@ -28,6 +29,52 @@
...
+    descriptor = os.open(
+        lock_path,
+        os.O_RDWR | os.O_CREAT | getattr(os, "O_CLOEXEC", 0),
+        0o600,
+    )
+    _memory_config_tx_state.depth = 1
+    try:
+        import fcntl
+
+        fcntl.flock(descriptor, fcntl.LOCK_EX)
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply config/v2_config.py]: Fixed: memory_config_transaction uses msvcrt on Windows and fcntl on POSIX, with CONFIG_LOCK acquired before the file lock.

File: core/memory/runtime.py [hunk 10]
```diff
@@ -1633,21 +1699,207 @@ async def _restart_locked(self) -> dict[str, Any]:
...
+            error = _runtime_error_for_status(
+                await asyncio.to_thread(self._artifact_manager.status)
+            )
+            self._runtime_error = error
+            return {"ok": False, "error": error, "result": "failed"}
+        rebuild_settings = _process_settings(
+            candidate,
+            call_log_db_path=self._call_log_db_path,
+        )
+        if not _rebuild_settings_usable(rebuild_settings):
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply core/memory/runtime.py]: Fixed: emptiness is inspected before requiring complete embedding credentials; empty roots settle without a rebuild child.

File: vibe/api.py [hunk 11]
```diff
@@ -929,7 +929,9 @@ def save_config(
     payload = _strip_preserved_config_secrets(payload)
     payload = _mark_explicit_audio_asr_enabled(payload)
 
-    with CONFIG_LOCK:
+    # Share the Memory candidate+marker transaction with Controller settlement so
+    # a whole-config writer cannot re-persist a stale embedding marker over it.
+    with memory_config_transaction():
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply vibe/api.py:934]: Fixed: lock order is CONFIG_LOCK then file lock so callers already holding CONFIG_LOCK cannot deadlock against Memory writers.

**Comment by cyhhao** [comment]
@codex review

File: vibe/api.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply vibe/api.py:934]: Fixed in 5397fd8e: `V2Config.save()` now always takes the Memory config transaction and, unless `persist_memory=True`, re-reads the durable Memory unit under that lock. Direct writers (language, Model Hub, agent auth, etc.) therefore cannot re-arm a settled marker or clobber concurrent Memory settlement. Memory-owning paths (`save_config(allow_memory=True)`, Controller settlement) pass `persist_memory=True`. Coverage: `test_direct_v2config_writer_preserves_settled_memory_marker`.

File: ui/src/components/settings/memory/MemorySettingsPanel.tsx [hunk 12]
```diff
@@ -150,81 +160,129 @@ export const MemorySettingsPanel: React.FC<{
...
-      }
       const res = await api.saveMemorySettings(patch);
       if (isMemoryOk(res)) {
         onSaved(res);
-        showToast(t('memory.settings.saved'), 'success');
+        const runtime = res.runtime as { ok?: boolean } | undefined;
+        if (runtime && typeof runtime.ok === 'boolean') {
+          // Only announce rebuild outcomes when the backend actually ran rebuild.
+          showToast(
+            runtime.ok ? t('memory.settings.rebuildCompleted') : t('memory.settings.rebuildFailed'),
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply ui/src/components/settings/memory/MemorySettingsPanel.tsx]: Fixed in 5397fd8e: rebuild completed/failed toasts are gated on `needsRebuild` (confirmed rebuild path). Ordinary reconcile results that only carry `runtime.ok` now show the normal saved toast. Coverage: MemorySettingsPanel tests for ordinary save vs confirmed rebuild.

File: vibe/ui_memory_routes.py [hunk 13]
```diff
@@ -298,23 +427,22 @@ async def _apply_memory_settings_patch(
                     api.save_memory_config,
                     rollback_payload,
                     embedding_change_pending=current.memory.embedding_change_pending,
+                    expected=saved.memory,
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply vibe/ui_memory_routes.py:431]: Fixed in 5397fd8e: when rollback hits `MemoryConfigStaleWrite` or reconcile reports `memory_operation_in_progress`, the route reloads the durable Memory unit and returns success if the requested operational fields still match (marker may have been settled by concurrent Retry). Coverage: `test_memory_enabled_save_succeeds_when_settlement_clears_marker`.

**Comment by cyhhao** [comment]
@codex review
